apps/builder/app/builder/features/address-bar.tsx

```diff
@@ -101,9 +133,166 @@ const useCopyUrl = (pageUrl: string) => {
   };
 };
 
-const AddressBar = forwardRef<HTMLFormElement, unknown>((_props, ref) => {
+const moveSelection = (menu: HTMLElement, diff: number) => {
+  const options = Array.from(menu.querySelectorAll("[role=option]"));
+  const index = options.findIndex((element) => element.ariaSelected === "true");
+  const newIndex = Math.max(-1, Math.min(index + diff, options.length - 1));
+  if (index >= 0) {
+    options[index].ariaSelected = null;
+  }
+  if (newIndex >= 0) {
+    options[newIndex].ariaSelected = "true";
+  }
+};
+
+/**
+ * Suggestions are opened whenever user
+ * - types in input
+ * - focuses input
+ * - press arrow down or arrow up
+ *
+ * and closed when
+ * - input is lost focus
+ * - escape or enter are pressed
+ *
+ * option selection is managed by arrow up, arrow down and hover
+ */
+const Suggestions = ({
+  containerRef,
+  options,
+  onSelect,
+}: {
+  containerRef: RefObject<HTMLFormElement>;
+  options: string[];
+  onSelect: (option: string) => void;
+}) => {
+  const list = options;
+
+  const menuRef = useRef<HTMLDivElement>(null);
+  const [isListOpen, setIsListOpen] = useState(false);
+
+  useEffect(() => {
+    const container = containerRef.current;
+    if (container === null) {
+      return;
+    }
+    const handleInput = () => {
+      setIsListOpen(true);
+    };
+    let frameId: undefined | number;
+    const handleFocusIn = () => {
+      if (frameId) {
+        cancelAnimationFrame(frameId);
+      }
+      setIsListOpen(true);
+    };
+    const handleFocusOut = () => {
+      frameId = requestAnimationFrame(() => {
+        setIsListOpen(false);
+      });
+    };
+    const handleKeyDown = (event: KeyboardEvent) => {
+      if (event.key === "ArrowDown") {
+        // avoid moving cursor to the end
+        event.preventDefault();
+        // trigger menu with up and down like in chrome
+        if (menuRef.current === null) {
+          setIsListOpen(true);
+          return;
+        }
+        moveSelection(menuRef.current, +1);
+      }
+      if (event.key === "ArrowUp") {
+        // avoid moving cursor to the start
+        event.preventDefault();
+        if (menuRef.current === null) {
+          setIsListOpen(true);
+          return;
+        }
+        moveSelection(menuRef.current, -1);
+      }
+      if (event.key === "Escape" && menuRef.current) {
+        // avoid closing popovers and dialogs when list is open
+        event.stopPropagation();
+        setIsListOpen(false);
+      }
+      if (event.key === "Enter" && menuRef.current) {
+        const selected = menuRef.current?.querySelector(
+          "[role=option][aria-selected=true]"
+        );
+        if (selected instanceof HTMLElement) {
+          // avoid submitting form when item is selected
+          event.preventDefault();
+          selected.click();
+        }
+      }
+    };
+    container.addEventListener("input", handleInput);
```

Comment:
was it impossible to use react for the handlers?

Reply:
yes, but would require using more refs instead of variable inside of effect

Reply:
feels like the ritght way to use react tbh, since it does its events delegation stuff and fixes things for some events